bundles/org.openhab.binding.heliosventilation/src/main/java/org/openhab/binding/heliosventilation/internal/HeliosVentilationHandler.java

```diff
@@ -0,0 +1,459 @@
+     * Update the variable corresponding to given channel/command
+     *
+     * @param channelUID UID of the channel to update
+     * @param command data element to write
+     *
+     */
+    public void update(ChannelUID channelUID, Command command) {
+        HeliosVentilationBindingConstants.DATAPOINTS.values().forEach((outer) -> {
+            HeliosVentilationDataPoint v = outer;
+            do {
+                if (channelUID.getThingUID().equals(thing.getUID()) && v.getName().equals(channelUID.getId())) {
+                    if (v.isWritable()) {
+                        byte[] txFrame = { 0x01, BUSMEMBER_ME, BUSMEMBER_CONTROLBOARDS, v.address(), 0x00, 0x00 };
+                        txFrame[4] = v.getTransmitDataFor((State) command);
+                        if (v.requiresReadModifyWrite()) {
+                            txFrame[4] |= memory.get(v.address()) & ~v.bitMask();
+                            memory.put(v.address(), txFrame[4]);
+                        }
+                        txFrame[5] = (byte) checksum(txFrame);
+                        tx(txFrame);
+
+                        txFrame[2] = BUSMEMBER_SLAVEBOARDS;
+                        txFrame[5] = (byte) checksum(txFrame);
+                        tx(txFrame);
+
+                        txFrame[2] = BUSMEMBER_MAINBOARD;
+                        txFrame[5] = (byte) checksum(txFrame);
+                        tx(txFrame);
+                    }
+                }
+                v = v.link();
+            } while (v != null);
+        });
+
+    }
+
+    /**
+     * calculate checksum of a frame
+     *
+     * @param frame filled with 5 bytes
+     * @return checksum of the first 5 bytes of frame
+     */
+    private int checksum(byte[] frame) {
+        int sum = 0;
+        for (int a = 0; a < 5; a++) {
+            sum += frame[a] & 0xff;
+        }
+        sum %= 256;
+        return sum;
+    }
+
+    /**
+     * interpret a frame, which is already validated to be in correct format with valid checksum
+     *
+     * @param frame 6 bytes long data with 0x01, sender, receiver, address, value, checksum
+     */
+    private void interpretFrame(byte[] frame) {
+        if ((frame[2] & BUSMEMBER_REC_MASK) == (BUSMEMBER_ME & BUSMEMBER_REC_MASK)) {
+            // something to read for us
+            byte var = frame[3];
+            byte val = frame[4];
+            if (HeliosVentilationBindingConstants.DATAPOINTS.containsKey(var)) {
+                HeliosVentilationDataPoint datapoint = HeliosVentilationBindingConstants.DATAPOINTS.get(var);
+                if (datapoint.requiresReadModifyWrite()) {
+                    memory.put(var, val);
+                }
+                do {
+                    if (logger.isTraceEnabled()) {
+                        String t = datapoint.asString(val);
+                        logger.trace("Received {} = {}", datapoint, t);
+                    }
+                    if (thing.getStatus() != ThingStatus.REMOVING) {
```

Comment:
I didn't put this for fun but think that I had an issue that serialEvent was still called in another thread before everything was closed, and updating state and status doesn't in this case doesn't make any sense.

Reply:
Interesting. Before the thing is removed, it should have completed `dispose()` and nothing should be alive then. If this happens, either this is a bug in the framework or something from the serial handling is still alive when `dispose()` finished.

Reply:
ok, I tried again and the need for this condition did not show up again, so I removed it now. I believe that one of the recently added synchronized keywords helped here or the openhab core changed since I initially wrote that some months ago.